operator/README.md

```diff
@@ -0,0 +1,173 @@
+
+#### DeploymentSpec
+
+|Field | type | Description | Default Value |
+|---|---|---|---|
+| driverName | string | Unique CSI driver name to use | `pmem-csi.intel.com` |
+| image | string | PMEM-CSI docker image name used for the deployment | `intel/pmem-csi-driver:canary` |
+| provisionerImage | string | [CSI provisioner](https://github.com/kubernetes-csi/external-provisioner) docker image name | `quay.io/k8scsi/csi-provisioner:v1.2.1` |
+| registrarImage | string | [CSI node driver registrar](https://github.com/kubernetes-csi/node-driver-registrar) docker image name | `quay.io/k8scsi/csi-node-driver-registrar:v1.1.0` |
+| pullPolicy | string | Docker image pull policy. either one of `Always`, `Never`, `IfNotPresent` | `IfNotPresent` |
+| logLevel | integer | PMEM-CSI driver logging level | 3 |
+| deviceMode | string | Device management mode to use. Supports one of `lvm` or `direct` | `lvm`
+| controllerResources | [ResourceRequirements](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#resourcerequirements-v1-core) | Describes the compute resource requirements for controller pod |
+| nodeResources | [ResourceRequirements](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#resourcerequirements-v1-core) | Describes the compute resource requirements for the pods running on node(s) |
+| nodeSelector | string map | [Labels to use for selecting Nodes](../README.md#run-pmem-csi-on-kubernetes) on which PMEM-CSI driver should run. | `{ "storage": "pmem" }`
+
+#### DeploymentStatus
+
+A PMEM-CSI Deployment's `status` field is a `DeploymentStatus` object, which has
+a `phase` field. The phase of a Deployment is high-level summary of where the
+Deployment is in it's lifecycle.
+
+The possible `phase` values and their meaning are as below:
+
+| value | meaning |
+|---|---|---|
+| Pending | The `Deployment` has been accepted by the system, but waiting for one or more of it's pre-requisites to be completed. Such as waiting for required [TLS certificates](../README.md#security) to be generated |
+| Initializing | The required sub-resources of the `Deployment` are being initialized |
+| Running | All the required sub-resources of the `Deployment` are created |
+| Failed | For some reason the state of the `Deployment` failed and cannot be progressed |
+
+### TLS Certificate approval
+
+All PMEM-CSI driver communication is protected by mutual TLS. So the driver
+deployment needs to generate those certificates. The operator uses Certificate
+Authority(CA) provided by the Kubernetes cluster to get those certificates. On a
+new deployment, the operator creates Kubernetes `CertificateSigningRequest`
+objects and keeps the deployment in `pending` state till those CSRs get approved
+by an authorized user so that Kubernetes CA could generate the certificates.
+Once the CSRs are approved and certificates are ready, then the deployment
+process gets progressed.
```

Comment:
"gets progressed" -> "proceeds"

Reply:
What happens if nodes are added later on that match the label for the PMEM-CSI daemon set?

It's not enough to create the CSRs once. This has to be a control loop, like everything else.

Reply:
The certificates are not generated per node. We create sets of certificates  `peme-registry` and `pmem-node-controller`. `pmem-node-controller` certificate is shared by all the nodes. So we do not generate a  new CSR upon a new pmem node.

Reply:
So that's the current status and it's kind of okayish for manual deployment. But do you agree that the operator must handle also dynamically added nodes?

Reply:
At least till now, I couldn't figure out something to do special for a newly added node. That should be handled by `daemonset` controller. Unless we think of implementing `auto labeling` new nodes feature.

Reply:
I'm not talking about doing something automatic *on* a node. Configuring PMEM and labeling the node is still something that the admin needs to do.

I'm talking about the situation where we need to do something *for* a node:
- PMEM-CSI operator is running.
- Admin creates a deployment.
- Deployment is running.
- Admin enables PMEM on a node and sets the label that is used by the deployment.

As it's designed right now, PMEM-CSI node driver will start to run on that node, but cannot communicate with the registry because it lacks the necessary certificates.

What we need is a control loop that reacts to label changes on nodes and then re-recreates the content of the `pmem-node-controller` secret.

Reply:
> As it's designed right now, PMEM-CSI node driver will start to run on that node, but cannot communicate with the registry because it lacks the necessary certificates.

We are not generating certificates per node(unlike the current [certificate generation scirpt](https://github.com/intel/pmem-csi/blob/devel/test/setup-ca-kubernetes.sh)), instead, all node divers share the same secret/certificate. So we might not need any extra step for setting up certificates for a newly added node, as the secret/certificate is already there and the Pod template of the daemonset has the reference to the existing secret.

Reply:
Oh, so it's a change in the security design. I wasn't aware of that. So it's a bit less secure than before (a rogue node can now impersonate any other node). Might still be good enough.

Reply:
Hmm, I might have misremembered here. I thought we had different certificates per node, but it seems that we don't. Perhaps that's what I had in OIM and PMEM-CSI always was a bit simpler - sorry for the confusion. :man_shrugging:

Reply:
Let's take a step back. Why are we creating different certificates for each node?

I believe the intention was to ensure that only a node controller which has the right certificate can register for the node. But that doesn't seem to be implemented (there's no certificate check in https://github.com/intel/pmem-csi/blob/599a94e3fca15384c385f674355cd177f49602f4/pkg/registryserver/registryserver.go#L93-L100) and therefore it now works for the operator where all nodes present the same certificate.

@avalluri: does that make sense? Why was strict certificate checking not implemented?

In OIM, I had implemented such additional checks, for example: https://github.com/intel/oim/blob/de83caab34dd9df2565ac2dcf7e1c164e79a3f2b/pkg/oim-registry/registry.go#L174-L184

Reply:
> I believe the intention was to ensure that only a node controller which has the right certificate can register for the node.

That was not the intention of the implementation though the script generating per node certificates.

> @avalluri: Why was strict certificate checking not implemented?

I felt like it's ok to not tiding the certificate to a node, instead the certificate _CommonName_ should match. Do you think we should have a more strict check?

Reply:
> That was not the intention of the implementation though the script generating per node certificates.

Then the script should have also generated only one certificate for all nodes. The way it is now is just confusing.

> I felt like it's ok to not tiding the certificate to a node, instead the certificate CommonName should match. Do you think we should have a more strict check?

I'm not sure. Probably we don't need it: if one node gets compromised, it can already disrupt the operation of PMEM-CSI also without impersonating other nodes.

So let's keep it simple, but that means that `setup-ca-kubernetes.sh` should also get simplified to match the actual implementation.